pcsx2-qt/Debugger/DebuggerSettingsManager.cpp

```diff
@@ -0,0 +1,163 @@
+}
+
+void DebuggerSettingsManager::writeJSONToPath(std::string path, QJsonDocument jsonDocument)
+{
+	QFile file(QString::fromStdString(path));
+	if (!file.open(QIODevice::WriteOnly))
+	{
+		Console.WriteLn(fmt::format("Debugger Settings Manager: Failed to write Debugger Settings file to path: '{}'", path));
+		return;
+	}
+	file.write(jsonDocument.toJson(QJsonDocument::Indented));
+	file.close();
+}
+
+void DebuggerSettingsManager::loadGameSettings(BreakpointModel* bpModel)
+{
+	const std::string path = VMManager::GetDebuggerSettingsFilePathForCurrentGame();
+	if (path == "")
+		return;
+
+	const QJsonValue breakpointsValue = loadGameSettingsJSON().value("Breakpoints");
+	const QString valueToLoad = breakpointsValue.toString();
+	if (breakpointsValue.isUndefined() || !breakpointsValue.isArray())
+	{
+		Console.WriteLn(fmt::format("Debugger Settings Manager: Failed to read Breakpoints array from settings file: '{}'", path));
+		return;
+	}
+
+	const QJsonArray breakpointsArray = breakpointsValue.toArray();
+	for (u32 row = 0; row < breakpointsArray.size(); row++)
+	{
+		const QJsonValue rowValue = breakpointsArray.at(row);
+		if (rowValue.isUndefined() || !rowValue.isObject())
+		{
+			Console.WriteLn("Debugger Settings Manager: Failed to load invalid Breakpoint object.");
+			continue;
+		}
+		const QJsonObject rowObject = rowValue.toObject();
+
+		QStringList fields;
+		u32 col = 0;
+		for (auto iter = rowObject.begin(); iter != rowObject.end(); iter++, col++)
+		{
+			QString headerColKey = bpModel->headerData(col, Qt::Horizontal, Qt::DisplayRole).toString();
+			//QString val = rowObject.value(headerColKey).toString();
+			fields << rowObject.value(headerColKey).toString();
+		}
+		bpModel->loadBreakpointFromFieldList(fields);
+	}
+}
+
+void DebuggerSettingsManager::loadGameSettings(SavedAddressesModel* savedAddressesModel)
+{
+	const std::string path = VMManager::GetDebuggerSettingsFilePathForCurrentGame();
+	if (path == "")
+		return;
+
+	const QJsonValue savedAddressesValue = loadGameSettingsJSON().value("SavedAddresses");
+	QString valueToLoad = savedAddressesValue.toString();
+	if (savedAddressesValue.isUndefined() || !savedAddressesValue.isArray())
+	{
+		Console.WriteLn(fmt::format("Debugger Settings Manager: Failed to read Saved Addresses array from settings file: '{}'", path));
+		return;
+	}
+
+	const QJsonArray breakpointsArray = savedAddressesValue.toArray();
+
+	for (u32 row = 0; row < breakpointsArray.size(); row++)
+	{
+		const QJsonValue rowValue = breakpointsArray.at(row);
+		if (rowValue.isUndefined() || !rowValue.isObject())
+		{
+			Console.WriteLn("Debugger Settings Manager: Failed to load invalid Breakpoint object.");
+			continue;
+		}
+		const QJsonObject rowObject = rowValue.toObject();
+		QStringList fields;
+		u32 col = 0;
+		for (auto iter = rowObject.begin(); iter != rowObject.end(); iter++, col++)
+		{
+			QString headerColKey = savedAddressesModel->headerData(col, Qt::Horizontal, Qt::DisplayRole).toString();
```

Comment:
It seems wrong to use the display name in the format here, especially since the headerData function is going to return a translated string. Additionally, I think it would be nice to have version number field in the format. What I typically do is just make it a single integer, then increment it whenever something changes.

Reply:
Good call, I had been meaning to add an `Export` role for the `SavedAddresses` model like `Breakpoints` has, but now need it for two things.

Can definitely add a version, but where are you referring? In the settings file as a whole (updates when anything changes) or per object (increments when saved addresses is updated, or breakpoints is updated individually)?

Reply:
For the version number I meant the file format version, so it would be a child of the root in the JSON file. So if you add a new feature to the format, or modify an existing one, you could increment the version number that would be hardcoded in the PCSX2 source code. That way if we modify the format in the future, we'll be able to make an attempt to still read old files by checking the version.

Reply:
It looks like Qt::UserRole is used while exporting now, but Qt::DisplayRole is still used when importing.

Reply:
Following up on this, `Qt::UserRole` is now used for all headers, and the models are updated to give the base string when retrieved with `Qt::UserRole` instead of a translated string.

I also added a version to the format, thought nothing is currently done with it, as I figure when changes are done, each importer can handle if separately, as for example Saved Addresses might be able to handle `0.00` still but BPs might have to skip `0.01` and warn for example.

Reply:
Should look something like this:

<img width="244" alt="image" src="https://github.com/PCSX2/pcsx2/assets/11288319/c1d5865b-7e6a-4dca-b784-04bea1780f8a">

Perhaps it's an access thing. I'll resolve it for you.